File: java/257.binary-tree-paths.java
/*
 * @lc app=leetcode id=257 lang=java
 *
 * [257] Binary Tree Paths
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public List<String> binaryTreePaths(TreeNode root) {
        List<String> res = new LinkedList<>();
        if (root != null) drawPaths(root, "", res);
        return res;
    }

    public void drawPaths(TreeNode root, String path, List<String> res) {
        if (root.left == null && root.right == null) res.add(path + root.val);
        if (root.left != null) drawPaths(root.left, path + root.val + "->", res);
        if (root.right != null) drawPaths(root.right, path + root.val + "->", res);
    }
}

File: go/257.binary-tree-paths.go
/*
 * @lc app=leetcode id=257 lang=golang
 *
 * [257] Binary Tree Paths
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func binaryTreePaths(root *TreeNode) []string {
    if root == nil {
		return []string{}
	}
	return bfs(root, "")
}

func bfs(root *TreeNode, path string) []string {
	if path == "" {
		path = strconv.Itoa(root.Val)
	} else {
		path = path + "->" + strconv.Itoa(root.Val)
	}
	if root.Left == nil && root.Right == nil {
		return []string{path}
	}
	res := []string{}
	if root.Left != nil {
		res = append(res, bfs(root.Left, path)...)
	}
	if root.Right != nil {
		res = append(res, bfs(root.Right, path)...)
	}
	return res
}
// @lc code=end

